goodai/ltm/memory.py

```diff
@@ -0,0 +1,169 @@
+import json
+import time
+from typing import Any
+from copy import deepcopy
+from collections import defaultdict
+from goodai.ltm.mem.auto import AutoTextMemory, DefaultTextMemory
+from goodai.ltm.mem.config import TextMemoryConfig
+from goodai.ltm.mem.base import RetrievedMemory, PassageInfo
+from multiprocessing import SimpleQueue, Process
+
+
+def memory_server(
+    mem_kwargs: dict, time_budget: float, query_queue: SimpleQueue, add_queue: SimpleQueue,
+    bkg_queue: SimpleQueue, processed_queue: SimpleQueue, out_queue: SimpleQueue,
+):
+    ltm = LTMSystem(**mem_kwargs)
+    while True:
+        t0_loop = time.time()
+        # Attend queries
+        mems = None
+        if not query_queue.empty():
+            d = query_queue.get()
+            assert d["method"] in {"retrieve", "retrieve_from_keywords"}
+            mems = getattr(ltm, d["method"])(**d["kwargs"])
+            out_queue.put(mems)
+        # Give additions and changes some slack. Otherwise, queries might hoard time.
+        while True:
+            t0_changes = time.time()
+            # Add memories and send it to background processing
+            # TODO: process these in random order or randomly drop some, to avoid bottlenecks
+            if not add_queue.empty():
+                kwargs = add_queue.get()
+                content, metadata = ltm.content_addition_preprocessing(**kwargs)
+                text_key = ltm.add_content(**kwargs)
+                bkg_kwargs = dict(text=content, metadata=metadata, text_key=text_key)
+                bkg_queue.put(bkg_kwargs)
```

Comment:
Unless I am missing something, this is not technically another process

Reply:
Actually I was looking at the wrong bit, I was looking at line 33.

Reply:
The addition of the text is nearly instantaneous, so it's done in the same process, but then the memory is forwarded to the background processing process. In this case, there are two main effects:
1. The main process doesn't have to wait for the addition, but just puts the data to the queue.
2. The memory server is freed from the post-processing, and can keep attending requests.

Reply:
ah, yeah... that part isn't very clean yet. I'm not sure how to properly do it without including the preprocessing in the server process and/or respecting the LTM's public interface.

Reply:
`ltm.content_addition_preprocessing()` is both an adaptor between the `LTMSystem` interface and the semantic memory's one, and a placeholder for any kind of pre-processing that should take place before adding the memory to the semantic database. It's there mainly because I need to access the `content` and `metadata` elements, whatever they are after the preprocessing.